src/execution.jl

```diff
@@ -138,21 +138,27 @@ macro roc(ex...)
     compiler_kwargs, call_kwargs = split_kwargs(kwargs)
     vars, var_exprs = assign_args!(code, args)
 
+    @gensym kernel_args kernel_tt device kernel queue signal
+    @show vars
+    @show var_exprs
+    @show compiler_kwargs
+    @show call_kwargs
+
     push!(code.args,
         quote
             GC.@preserve $(vars...) begin
-                local kernel_args = map(rocconvert, ($(var_exprs...),))
-                local kernel_tt = Tuple{Core.Typeof.(kernel_args)...}
-                local device = extract_device(; $(map(esc, call_kwargs)...))
-                local kernel = rocfunction(device, $(esc(f)), kernel_tt;
-                                           $(map(esc, compiler_kwargs)...))
-                local queue = extract_queue(device; $(map(esc, call_kwargs)...))
-                local signal = create_event()
-                kernel(queue, signal, kernel_args...; $(map(esc, call_kwargs)...))
-                wait(signal)
+                local $kernel_args = map(rocconvert, ($(var_exprs...),))
+                local $kernel_tt = Tuple{Core.Typeof.($kernel_args)...}
+                local $device = AMDGPUnative.extract_device(; $(call_kwargs...))
```

Comment:
You can also do `$extract_device` etc to resolve the function eagerly.

Reply:
Will push this with my next commit.